### PR TITLE
Switched engine to use buffer collection.

### PR DESCRIPTION
**Pull request by dragostis**
The old way of allocating images meant that one would have to make sure that Scenic and Flutter were using exactly the same pixel formats. This patch removes the old image allocation and replaces it with a sysmem API that uses buffer collections instead. This permits a smooth negotiation of formats between the two systems.

### CODE REVIEW

**Review by arbreng — APPROVED** [review]
This looks fantastic!

Please fix the failing checks (or wait for them to be deflaked) and then we can merge this ASAP

**Comment by chinmaygarde** [comment]
The patch needs to be rebased and the licenses fixed to proceed.

File: shell/platform/fuchsia/flutter/vulkan_surface.cc
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface.cc:47]: can we use __builtin_ctz instead of adding this?

File: shell/platform/fuchsia/flutter/vulkan_surface.cc
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface.cc:250]: I don't think we should connect to sysmem each time we allocate an image. Can this connection live in the surface producer and be reused for all allocations?

File: shell/platform/fuchsia/flutter/vulkan_surface.h
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface.h:14]: nit: remove if not needed

File: shell/platform/fuchsia/flutter/vulkan_surface.cc
Comment (arbreng) [reply shell/platform/fuchsia/flutter/vulkan_surface.cc]: +1, it is inefficient to do this every time we allocate an image

File: shell/platform/fuchsia/flutter/vulkan_surface.cc
Comment (arbreng) [reply shell/platform/fuchsia/flutter/vulkan_surface.cc]: +1, I checked and both GCC and clang have this builtin

File: shell/platform/fuchsia/flutter/vulkan_surface_pool.h
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface_pool.h:45]: what's this line for?

File: shell/platform/fuchsia/flutter/vulkan_surface.cc
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface.cc:269]: Do we need these? It used to be that we needed it to match scenic but now that we're using sysmem it's enough to set the usage flutter needs. This would also require changing line 74.

File: shell/platform/fuchsia/flutter/vulkan_surface.cc
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface.cc:43]: Can we remove this ifdef now?

File: shell/platform/fuchsia/flutter/vulkan_surface.cc
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface.cc:8]: is this still needed?

File: shell/platform/fuchsia/flutter/vulkan_surface.cc
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface.cc:291]: why is this needed? failing to see when tiling will ever be set to LINEAR and even then I'm not sure why we want this..

File: shell/platform/fuchsia/flutter/vulkan_surface.h
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface.h:177]: nit: const uint32_t as set by ctor

**Review by dreveman — CHANGES_REQUESTED** [review]

File: shell/platform/fuchsia/flutter/vulkan_surface.cc
Comment (dragostis) [reply shell/platform/fuchsia/flutter/vulkan_surface.cc:269]: What usage does Flutter need?

File: shell/platform/fuchsia/flutter/vulkan_surface.cc
Comment (dragostis) [reply shell/platform/fuchsia/flutter/vulkan_surface.cc:43]: We should keep the else branch, right?

File: shell/platform/fuchsia/flutter/vulkan_surface.cc
Comment (dreveman) [reply shell/platform/fuchsia/flutter/vulkan_surface.cc:268]: Don't know. We used to need all these to match the usage in Scenic but that's no longer needed when using sysmem. Let's keep them as is but put a comment here about removing usages that are not required by Flutter.

File: shell/platform/fuchsia/flutter/vulkan_surface.cc
Comment (dreveman) [reply shell/platform/fuchsia/flutter/vulkan_surface.cc:42]: No, else branch is what we want to remove. If branch is required for AFBC so critical to keep. The else branch was here to match intel image import details in Scenic but sysmem should remove the need for this.

File: shell/platform/fuchsia/flutter/vulkan_surface.cc
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface.cc:376]: I noticed that this check was removed. I assume that's intentional?

File: shell/platform/fuchsia/flutter/vulkan_surface.cc
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface.cc:311]: - why is the "memory_type == 0" check needed? if needed, why not "0 ? 0 : ... "
- nit: static_cast instead of c-style cast

File: shell/platform/fuchsia/flutter/vulkan_surface.cc
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface.cc:67]: nit: start comment with FIXME or TODO so it's clear that it should be improved by follow up change
nit: can you also move these to a constant at the top of the file so it's easier to see that sysmem and vulkan usages should both be adjusted?

File: shell/platform/fuchsia/flutter/vulkan_surface.cc
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface.cc:261]: nit: move this next to the constant for vulkan usages so it's less likely that we update just one of them

**Review by dreveman — APPROVED** [review]
lgtm after addressing this last set of comments

File: shell/platform/fuchsia/flutter/vulkan_surface.cc
Comment (dragostis) [reply shell/platform/fuchsia/flutter/vulkan_surface.cc:311]: I don't think `memoryTypeBits` are guaranteed not to be `0` and `__builtin_ctz` has UB for `0`. I'll change this to an `FML_DCHECK`.

File: shell/platform/fuchsia/flutter/vulkan_surface.cc
Comment (dragostis) [reply shell/platform/fuchsia/flutter/vulkan_surface.cc:376]: This check doesn't seem to be needed anymore since the only place where `PushSessionImageSetupOps` is in the constructor where `sk_surface_` being `nullptr` would return before this can execute.

File: shell/platform/fuchsia/flutter/vulkan_surface.cc
Comment (dreveman) [reply shell/platform/fuchsia/flutter/vulkan_surface.cc]: If memoryTypeBits is 0 then something is very wrong and we can't allocate memory for the image so DCHECK makes sense.

**Pull request merged**
